crates/ra_hir/src/ty/infer.rs

```diff
@@ -481,10 +496,50 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
 
         match resolved {
             Resolution::Def(def) => {
+                // Helpper function for finding self types
+                let find_self_types = || -> Option<Substs> {
```

Comment:
it would be nice to extract this into its own method :) it should just need to take the `actual_def_ty` and the `impl_block`.